### PR TITLE
Fix diffs of files that have quoted paths

### PR DESCRIPTION
**Pull request by jcouball**
### Your checklist for this pull request
🚨Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

- [X] Ensure all commits include DCO sign-off.
- [X] Ensure that your contributions pass unit testing.
- [X] Ensure that your contributions contain documentation if applicable.

### Description
As was reported in #418, if `git diff` reports a file path that has special characters like unicode, backslash, double quotes, etc. then the diff fails.

#### lib/git/diff.rb
The problem was found to be in the diff parser not knowing how to deal with quoted paths. This PR changes Git's behavior to always quote paths that contain special characters and escape those special characters with backslashes in the same way C escapes control characters (e.g. \t for TAB, \n for LF, \\ for backslash) or bytes with values larger than 0x80 (e.g. octal \302\265 for "micro" in UTF-8).

This Diff class uses the new `Git::EscapedPath` class to unescape a path that conforms to these rules.

#### lib/git/lib.rb
For every command, set the core.quotePath configuration setting to true by adding `-c core.quotePath=true` to any git command that is executed. See the details of this option in the [core.quotePath documentation](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corequotePath).

The encoding methods were moved to Git::EncodingUtils

#### lib/git/encoding_utils.rb
This module gathers all the encoding methods that were in Git::Lib so they can be shared between different classes.

#### lib/git/escaped_path.rb
The Gt::EscapedPath class represents an escaped Git path string and provides a method to unescape the string.

#### tests/units/test_logger.rb

Since the `-c core.quotePath` was added to every command line, the logger tests had to be changed to ignore changes in global options.

### Other

Some characters (like a double quote I learned) are not allowed in Windows filenames.  A complete enumeration of the rules for Windows filenames can be found in the article [Naming Files, Paths, and Namespaces](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file)

### CODE REVIEW

File: lib/git/diff.rb [hunk 1]
```diff
@@ -129,8 +129,8 @@ def process_full_diff
         final = {}
         current_file = nil
         @full_diff.split("\n").each do |line|
-          if m = /^diff --git a\/(.*?) b\/(.*?)/.match(line)
-            current_file = m[1]
+          if m = %r{\Adiff --git ("?)a/(.+?)\1 \1b/(.+?)\1\z}.match(line)
```
Comment (wasanx25) [inline lib/git/diff.rb]: It not work if double quotes begin with b. 

```suggestion
          if m = %r{\Adiff --git ("?)a/(.+?)\1 \1("?)b/(.+?)\1\z}.match(line)
```

```console
$ ruby -v
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
$ irb
irb(main):001:0> pattern_a = 'diff --git "a/asdf\"asdf" "b/asdf\"asdf"'
irb(main):002:0> pattern_b = 'diff --git a/asdfasdf "b/asdf\"asdf"'
irb(main):003:0> old_regex = %r{\Adiff --git ("?)a/(.+?)\1 \1b/(.+?)\1\z}
irb(main):004:0> new_regex = %r{\Adiff --git ("?)a/(.+?)\1 \1("?)b/(.+?)\1\z}
irb(main):005:0> old_regex.match(pattern_a)
=> #<MatchData "diff --git \"a/asdf\\\"asdf\" \"b/asdf\\\"asdf\"" 1:"\"" 2:"asdf\\\"asdf" 3:"asdf\\\"asdf">
irb(main):006:0> old_regex.match(pattern_b)
=> nil
irb(main):007:0> new_regex.match(pattern_a)
=> #<MatchData "diff --git \"a/asdf\\\"asdf\" \"b/asdf\\\"asdf\"" 1:"\"" 2:"asdf\\\"asdf" 3:"" 4:"asdf\\\"asdf">
irb(main):008:0> new_regex.match(pattern_b)
=> #<MatchData "diff --git a/asdfasdf \"b/asdf\\\"asdf\"" 1:"" 2:"asdfasdf" 3:"\"" 4:"asdf\\\"asdf\"">
```

**Review by wasanx25 — COMMENTED** [review]
It's great PR. I have same issue when Japanese file name rename.

**Comment by wasanx25** [comment]
@jcouball Do you plan to merge this pull request? I am having problem yet.

File: lib/git/diff.rb (hunk 1 above)
Comment (jcouball) [reply lib/git/diff.rb]: @wasanx25 I have changed the implementation to accommodate your problem where either a/ or b/ might be quoted but not both. The regex I went with is a little different than what you suggested:

```Ruby
if m = %r{\Adiff --git ("?)a/(.+?)\1 ("?)b/(.+?)\3\z}.match(line)
```

**Comment by jcouball** [comment]
@wasanx25 are you able to test this with your use case. I know it has been a long time.

**Comment by wasanx25** [comment]
@jcouball Sure, try do test but I need to remember use case so take a while.

**Comment by wasanx25** [comment]
@jcouball It passed test of my use case. Rename Japanese filename `日本語ファイル.txt` to `2日本語ファイル.txt`.

exec.rb
```ruby
require 'git'

g = Git.open('./')

commit1 = 'f66a6371dce97e5212bb5d7a61dbc81a8023eb53'
commit2 = 'e4c26f70168aced95beeb33158767bf5d86261f4'

puts g.diff(commit1, commit2).map(&:path)
```


```console
$ git diff f66a6371dce97e5212bb5d7a61dbc81a8023eb53 e4c26f70168aced95beeb33158767bf5d86261f4 | cat
diff --git "a/\346\227\245\346\234\254\350\252\236\343\203\225\343\202\241\343\202\244\343\203\253.txt" "b/2\346\227\245\346\234\254\350\252\236\343\203\225\343\202\241\343\202\244\343\203\253.txt"
similarity index 100%
rename from "\346\227\245\346\234\254\350\252\236\343\203\225\343\202\241\343\202\244\343\203\253.txt"
rename to "2\346\227\245\346\234\254\350\252\236\343\203\225\343\202\241\343\202\244\343\203\253.txt"

$ ruby exec.rb
日本語ファイル.txt
```

use git-1.10.0
```console
$ ruby exec.rb
/Users/wasanx25/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/git-1.10.0/lib/git/diff.rb:148:in `block in process_full_diff': undefined method `[]' for nil:NilClass (NoMethodError)
        from /Users/wasanx25/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/git-1.10.0/lib/git/diff.rb:131:in `each'
        from /Users/wasanx25/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/git-1.10.0/lib/git/diff.rb:131:in `process_full_diff'
        from /Users/wasanx25/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/git-1.10.0/lib/git/diff.rb:110:in `process_full'
        from /Users/wasanx25/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/git-1.10.0/lib/git/diff.rb:68:in `each'
        from exec.rb:8:in `map'
        from exec.rb:8:in `<main>'
```

Thanks,

**Pull request merged**